### ISSUE TITLE
Customizing FPS for Win32Platform

### ISSUE THREAD

**Issue by MeltyPlayer** [AvaloniaUI/Avalonia#16417]
### Is your feature request related to a problem? Please describe.

Currently, Win32Platform is hardcoded to run at 60fps here:
https://github.com/AvaloniaUI/Avalonia/blob/c47fd63e6246789d95bcff626746c7ed63f96e35/src/Windows/Avalonia.Win32/Win32Platform.cs#L88

### Describe the solution you'd like

LinuxFramebufferPlatform currently supports customizable FPS:
https://github.com/AvaloniaUI/Avalonia/blob/c47fd63e6246789d95bcff626746c7ed63f96e35/src/Linux/Avalonia.LinuxFramebuffer/LinuxFramebufferPlatformOptions.cs#L12
https://github.com/AvaloniaUI/Avalonia/blob/c47fd63e6246789d95bcff626746c7ed63f96e35/src/Linux/Avalonia.LinuxFramebuffer/LinuxFramebufferPlatform.cs#L65-L67

It'd be nice if Win32Platform had this same support, primarily to support high-framerate displays (i.e. 120+ fps) but also to potentially cap at lower framerates for lower spec hardware (i.e. 30 fps).

### Additional context

Given that this appears to have been a very tiny change for Linux, it seems like this would be very simple to implement. Given that this hasn't been added yet, I wanted to reach out to ask if there were any concerns (i.e. Avalonia isn't optimized enough yet, or is buggy at higher framerates)?

**Comment by rabbitism** [comment]
Related https://github.com/AvaloniaUI/Avalonia/issues/14073#issuecomment-1873488557

**Comment by maxkatz6** [comment]
> Currently, Win32Platform is hardcoded to run at 60fps here:

It's not. [It uses display VSync by default](https://github.com/AvaloniaUI/Avalonia/blob/c47fd63e6246789d95bcff626746c7ed63f96e35/src/Windows/Avalonia.Win32/WinRT/Composition/WinUiCompositorConnection.cs#L90-L93), where you can't change framerate, it can go far higher than 60fps normally. You can disable it in CompositionMode option, but it's not advised.
Adding FPS option might be confusing, as changing its value won't affect majority of Avalonia apps.

60 is only forced when other render timers are used, like a fallback for older systems or when ShouldRenderOnUIThread is used.

> LinuxFramebufferPlatform currently supports customizable FPS:

Framebuffer doesn't have any vsync, so that's the only option there, yes.

**Comment by MeltyPlayer** [comment]
> > Currently, Win32Platform is hardcoded to run at 60fps here:
> 
> It's not. [It uses display VSync by default](https://github.com/AvaloniaUI/Avalonia/blob/c47fd63e6246789d95bcff626746c7ed63f96e35/src/Windows/Avalonia.Win32/WinRT/Composition/WinUiCompositorConnection.cs#L90-L93), where you can't change framerate, it can go far higher than 60fps normally. You can disable it in CompositionMode option, but it's not advised. Adding FPS option might be confusing, as changing its value won't affect majority of Avalonia apps.
> 
> 60 is only forced when other render timers are used, like a fallback for older systems or when ShouldRenderOnUIThread is used.
> 
> > LinuxFramebufferPlatform currently supports customizable FPS:
> 
> Framebuffer doesn't have any vsync, so that's the only option there, yes.

Ah, thanks for the clarification!

Based on my experience, it really did feel like Avalonia was locked at 60 fps, since even the smallest possible OpenGL app (i.e. just clearing the screen) would only run at 60 fps. However, I tried out the Avalonia sample projects and noticed those were indeed running at 120 fps.

I've narrowed down the issue I was seeing to the following code in the `AppBuilder`:
```
.With(new Win32PlatformOptions {
    RenderingMode = [Win32RenderingMode.Wgl],
})
```

When the RenderingMode is set to Win32RenderingMode.Wgl, the highest my app can go is 60 fps; when this is removed, it can go up to 120 fps. It looks like you flagged this issue actually in a different issue, I just didn't hadn't noticed that when I was first investigating this.
https://github.com/AvaloniaUI/Avalonia/discussions/13125#discussioncomment-7205156

**Comment by maxkatz6** [comment]
Yeah, you are correct. Wgl won't support CompositionMode vsync. We might need to use wglSwapIntervalEXT in order to support similar behavior there. If not, it's better to add FPS config selector to new `WglOptions` class or similar.

Btw, why do you need WGL? Normally, we don't recommend it, unless it's necessary for your dependencies.

**Comment by MeltyPlayer** [comment]
Ah, sounds good! Yes, it doesn't necessarily need to be in Win32PlatformOptions, just wherever would be best for Avalonia in the long term.

My Avalonia app is a model viewer that leverages OpenTK for rendering a 3D view, I've hooked into Avalonia's OpenGL logic using an approach similar to this library:
https://github.com/SamboyCoding/OpenTKAvalonia

I haven't done a comprehensive investigation so I'm not sure if Wgl is 100% necessary, but I get memory access violations when I try any other Win32RenderingMode.

**Comment by maxkatz6** [comment]
> Note that, in order for events to be called on Windows, Avalonia must be able to load the OpenGL bindings, which requires switching the default rendering mode.

Is it still true with Angle OpenGL though, which is default? Normal OpenGlControlBase works there just fine.

**Comment by kekekeks** [comment]
WGL isn't really supported as the primary backend. If you need to present non-ANGLE OpenGL rendering, use WGL_NV_DX_interop, ID3D11Texture and https://github.com/AvaloniaUI/Avalonia/issues/9925

**Comment by MeltyPlayer** [comment]
Just following up on this, I have found a reason why Wgl is better for my use case: latency when setting up shaders. 

I'm creating a UI that supports loading 3d models/scenes, so I have to compile a lot of dynamic shaders whenever a new file is opened. What loads instantaneously in Wgl takes at least a second for AngleEgl, and files that load slower in Wgl can be 5-10s in the AngleEgl version. I've optimized everything pretty extensively but I have not been able to improve Angle performance down to anything less than a second.

I'm assuming this overhead is just expected when using AngleEgl? It's a bit too painful for my UI, unfortunately.

I'll try seeing if I can use the suggestions from https://github.com/AvaloniaUI/Avalonia/issues/16417#issuecomment-2251965747 to enable 120fps in Wgl instead.

**Comment by MeltyPlayer** [comment]
Just following up on this one last time, I found another issue where interop was mentioned:
https://github.com/AvaloniaUI/Avalonia/discussions/19061

Thankfully they had an example! Basically, this sets up Avalonia to use AngleEgl (which supports 120fps) but then circumvents that and tries to use Wgl/DX interop instead:
https://github.com/Dragorn421/DragoStuff/blob/aa1ac3434f2701739570adf77377c29e1e3171c1/SharpDXInteropControl.cs

I used this as a starting point but optimized it a bit so GL/interop stuff is only initialized once rather than each frame. This was enough to get up to a mostly stable 120fps, but I'm sure there's plenty more to optimize here:
https://github.com/MeltyPlayer/MeltyTool/blob/c5c7d2a3be9aa5d72450d2add46415945ca5d767/FinModelUtility/Fin/Fin.Ui.Avalonia/gl/SharpDxInteropControl.cs

Hopefully this helps anyone else who runs into this in the future! 

If you guys on the Avalonia team never intend to support 120fps for Wgl, just feel free to close this out; this not really an active issue for me now that I have a workaround.

**Comment by T1ti** [comment]
Is there a way to have a decent opengl renderer with desktop opengl (not ES) on all platforms ? the DX solution is windows only, and WGL is scuffed.
